### ISSUE TITLE
Allow wider types for request body

### ISSUE THREAD

**Issue by gustavoguichard** [gustavoguichard/make-service#28]
Our current `EnhancedRequestInit`'s `body` accepts: `body?: JSONValue | BodyInit | null` and `JSONValue` is:
```ts
type JSONValue =
  | string
  | number
  | boolean
  | { [x: string]: JSONValue }
  | Array<JSONValue>
```

I've done that to simulate the correct network possibilities as you can't send a (e.g.) function through the network.
The problem is that we are used to send things such as `Date` through the network knowing it will be `stringified`.

That's the dilemma: We want to enforce only valid JSON values but it is bad DX for the developer to send a body with a (e.g.) date property even though that will be stringified and we are all used to that.
Should we widen the type of `JSONValue` to include dates or should we just accept any `{ [x: string]: unknown }` and let the `JSON.stringify` do its magic without bothering?

Any opinions @danielweinmann and @diogob ?

```ts
service.post('/users', {
  body: {
    name: "John Doe",
    // this is currently invalid
    createdAt: new Date(),
  }
})
```

The code above is currently invalid but it creates a bad DX and enforcing correct JSON values at type level doesn't seem to be an advantage here.

**Comment by diogob** [comment]
Unfortunatelly the `JSON.stringify` function will generate a lot of surprising behaviour.
For instance:
```ts
JSON.stringify(new Date()) 
```
evaluates to `"2023-06-28T00:03:04.082Z"` which makes a lot of sense. However
```ts
JSON.stringify(() => new Date()) 
```
evaluates to `undefined`. Which is another way of saying that the function is partial and has no value defined for functions. Having said that I think it's good DX to ban partial functions on the type level.

Therefore my vote goes to include dates to the type `JSONValue`. Which could be renamed to `JSONable` for the sake of accuracy :stuck_out_tongue:

**Comment by gustavoguichard** [comment]
@diogob , thanks for the comment!
What do you think about copying [Remix's `Serialize`](https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/serialize.ts#L25-L36)?

Check [this example](https://www.typescriptlang.org/play?#code/PTAEDMHsCcGMFNQBUCeAHeBnW0CWaAXUAyUTXAWzQBtdwViALRA9eAKFY1AAVp4CrADxIAfKAC8oAN6gA2gGlQuAHagA1vBSRwyALoAuZIr2gAvqABkMswG5ObXv0Eok0AIYrMUaBREAaUABVcSk5JFN4AA8CeBUAE0x5INMAfmRQIz4BYRD7B24AKUxIFT5KXAJcADdECXZQUAAfMgI8FQBzBubQFQBXCgAjeGhulsHISGp4TzHQAGU21S7GloA5AeHR1dAAIUnp2Z3+6mp7LkQ10uLS8opKmrrQPoT4cFV4eJ6AMRfYKtKPUwKCGU3ywAAVA0IaAAAaeFCw5RJAjMUClagMC5MdxEWCeUDTIjaPqgeAARz6uMQsIADEiAO6VRhwgCMsOhC3giEYgjQmAMIEwBHcsHUkFq0HA1EgDIAdLBIBRgO5gAAWACcGoATAAOADMADZacBWZrabSNazocACogAJKYACCKhQIlCoFpZJicUSoFZVgy6TafUQRnA7momHg4LAaGcVRGAFpcB0VDAONj5iNcJHcAAveDuyTdR0ut1ib2xBIo6Ch0DpBGZbpIKu+pI3Mp4e5VWoNjIGFttmugK4qTt3B599IqeCS5uNVvRat+2QkQrzADyawAFABKIyqcAjYLmftBBcZZft+RpW+Xpc+kdyF7qdMMlSBOXf1-vlRyPQ72yFw3E8bwYD8JBAmzPA80LJA+hoIsxHEQdF2HP0ACUZniDEUCdaAPDdI8TxCfsdwva8RzHCduynRAZz6U5Mi5WDaELIQQj3ACHwwpJIEGAAreB-n7YDWFArwfEg6Cczg+BNyEkSCE4153lneIkEgTdCFwUpI3dURUO6WdJXYcEIRhDdtzIOT2MQOQCEQ6ZMD0HdeQIflBWABlfLlC5sDwQhqE8Do5RgDpgFw2BMGARhPHiCZIHUYBtWAAThP+TA5V5ChqAAYicpCkwCvdQAhW0szsgt4AQpCRD454VDfWV-z0D1HxXJI5BI6BQG+L9v160BML0bp0jkGDc3soRvlEQa5Sm+S6umIRMNEMbGiMACLKsrc1lstiavRJSstAHd0yIVRYGoPp4mWUB3EI9wUEwMqEtAG73EwaMkgqu1WOmmrFMylTOpvDLlI9aRujkdRlDUTRtF0VtvuMdRIifP0aJKLsKl7BjejnE8jAxowlpmpA4fa+w7HYdhIU5b4YEetRIdEqDlF0SpQHipIm3jSAMGgRMUXiognsQdwmr0tQmVROEXniN4PniWF-E5Ch3E0JhIGjBGSHRXT9OoUBBeF0WEeFHC5WhTkAFEoncKhpiMWR3CMYV2g6HolZVjSLCTJNxHd1JPaWTozBtAGgjU1WtJ0gEVAM8GR3Zghoe6EAnCFkZWGRXpIAlmXAXlllYT99TPnVjRuTQR7MBTTBYfh1Ra+RjI0cr1XGqpjH+1MkmNEMdG9FpwMYcabO+FzkWGAAfVwTB58eku5eZOFlartW5S5CX+ON5PTZ3apczhVJYXehJQH4CgJRpbuNKRcBoCVJhEBeWWW4R9udE7pJH6fF7tTfs8MjCD2gHoMOoBHY3TukWboi5qYa0aI0QB8RuiiFpvkAGSAsAEApsdKQhCOKyGThQeARgABEABxPowp3DVEgFQwI4AVBGH3JIcQTDcDxFYZMIwAARakZ5rCyEGE9IwAYzCiCAA) where I stripped some of its content (related to the `defer` feature).

What do you think?

**Comment by diogob** [comment]
I think it's not so much what we expect on the other side of the network, but about the expected behaviour of our request.

As a developer that does not fully understand JSON serialization, I would prefer to be told by the type system that I'm trying to send something that will be silently ignored.

Perhaps we should focus on what are the cases where the current typing breaks? 
This issue started with `Date`, are there any other examples that come to mind?

The best approach might be to just add `Date` for now and defer a more general type for the future once more cases become apparent.

At least those are my two cents :smiley:

**Comment by garusis** [comment]
I tend to agree with @diogob about this topic. I think it could be very early to start thinking in broader types, and it is always better for a developer to know what is or is not serializable. I think adding `Date` is good enough for now, and adding new types in the future only if it is possible to confirm that they are serializable, and maybe in the future, if it requires it, making something more flexible.

**Comment by gustavoguichard** [comment]
I'm sold! If anyone wants to make that PR you're welcome. Otherwise I'll probably be doing it next week ;)

**Comment by gustavoguichard** [comment]
Solved in #34
